clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -458,32 +458,38 @@ public ReadyCheckResult ready(Cluster cluster, long nowMs) {
 
         boolean exhausted = this.free.queued() > 0;
         for (Map.Entry<TopicPartition, Deque<ProducerBatch>> entry : this.batches.entrySet()) {
-            TopicPartition part = entry.getKey();
             Deque<ProducerBatch> deque = entry.getValue();
-
-            Node leader = cluster.leaderFor(part);
+            boolean empty;
             synchronized (deque) {
-                if (leader == null && !deque.isEmpty()) {
+                empty = deque.isEmpty();
```

Comment:
I'm not sure if this lock is worth taking in a fine grained way. I'm pretty sure it's safe, but may be over optimizing. The main problem is that cluster.leaderFor was outside of the lock previously, and I wanted to avoid calling from within the lock, given that it can be expensive.